docker/container.go

```diff
@@ -126,144 +117,64 @@ func name(names []string) string {
-			"name": c.Name(),
-		})
-	}
-
-	return container, err
+// Remove removes the container.
+func (c *Container) Remove(ctx context.Context, removeVolume bool) error {
+	return c.client.ContainerRemove(ctx, c.container.ID, types.ContainerRemoveOptions{
+		Force:         true,
+		RemoveVolumes: removeVolume,
+	})
 }
 
 // Stop stops the container.
 func (c *Container) Stop(ctx context.Context, timeout int) error {
-	return c.withContainer(ctx, func(container *types.ContainerJSON) error {
-		return c.client.ContainerStop(ctx, container.ID, time.Duration(timeout)*time.Second)
-	})
+	return c.client.ContainerStop(ctx, c.container.ID, time.Duration(timeout)*time.Second)
 }
 
 // Pause pauses the container. If the containers are already paused, don't fail.
 func (c *Container) Pause(ctx context.Context) error {
-	return c.withContainer(ctx, func(container *types.ContainerJSON) error {
-		if !container.State.Paused {
-			return c.client.ContainerPause(ctx, container.ID)
+	if !c.container.State.Paused {
+		if err := c.client.ContainerPause(ctx, c.container.ID); err != nil {
+			return err
 		}
-		return nil
-	})
+		return c.updateInnerContainer(ctx)
+	}
+	return nil
 }
 
 // Unpause unpauses the container. If the containers are not paused, don't fail.
 func (c *Container) Unpause(ctx context.Context) error {
-	return c.withContainer(ctx, func(container *types.ContainerJSON) error {
-		if container.State.Paused {
-			return c.client.ContainerUnpause(ctx, container.ID)
+	if c.container.State.Paused {
+		if err := c.client.ContainerUnpause(ctx, c.container.ID); err != nil {
+			return err
 		}
-		return nil
-	})
-}
-
-// Kill kill the container.
-func (c *Container) Kill(ctx context.Context, signal string) error {
-	return c.withContainer(ctx, func(container *types.ContainerJSON) error {
-		return c.client.ContainerKill(ctx, container.ID, signal)
-	})
-}
-
-// Delete removes the container if existing. If the container is running, it tries
-// to stop it first.
-func (c *Container) Delete(ctx context.Context, removeVolume bool) error {
-	container, err := c.findExisting(ctx)
-	if err != nil || container == nil {
-		return err
+		return c.updateInnerContainer(ctx)
 	}
+	return nil
+}
 
-	info, err := c.client.ContainerInspect(ctx, container.ID)
+func (c *Container) updateInnerContainer(ctx context.Context) error {
+	container, err := GetContainer(ctx, c.client, c.container.ID)
 	if err != nil {
 		return err
 	}
-
-	if !info.State.Running {
-		return c.client.ContainerRemove(ctx, container.ID, types.ContainerRemoveOptions{
-			Force:         true,
-			RemoveVolumes: removeVolume,
-		})
-	}
-
+	c.container = container
 	return nil
 }
 
+// Kill kill the container.
+func (c *Container) Kill(ctx context.Context, signal string) error {
+	return c.client.ContainerKill(ctx, c.container.ID, signal)
+}
+
 // IsRunning returns the running state of the container.
+// FIXME(vdemeester): remove the nil error here
```

Comment:
keeping the signature for now to not have too much noise on the diff, but will definitely do after.